k8s/upgrade/src/upgrade_resources/upgrade.rs

```diff
@@ -736,3 +748,69 @@ pub async fn upgrade_job_completed(
     }
     Ok(false)
 }
+
+struct ImageProperties {
+    pull_secrets: Option<Vec<k8s_openapi::api::core::v1::LocalObjectReference>>,
+    registry: String,
+    pull_policy: Option<String>,
+}
+
+impl From<Deployment> for ImageProperties {
+    fn from(d: Deployment) -> Self {
+        if d.spec.is_none() {
```

Comment:
let's a lot of unwraps in here...
can be handled by doing things like:
```
let Some(spec) = d.spec else { eprintln!("Failed to find .spec in Deployment.");
            std::process::exit(1); };
```
But instead of exit, you should use TryFrom and return an error